### DISCUSSION TITLE
Overview of supported APIs and customization needed

### DISCUSSION THREAD

**Discussion by ChristianGalla** [SharePoint/sp-dev-docs#9181] [General]
### What type of issue is this?

other

### What SharePoint development model, framework, SDK or API is this about?

other (enter in the "Additional environment details" area below)

### Target SharePoint environment

other (enter in the "Additional environment details" area below)

### What browser(s) / client(s) have you tested

- [ ] 💥 Internet Explorer
- [ ] 💥 Microsoft Edge
- [ ] 💥 Google Chrome
- [ ] 💥 FireFox
- [ ] 💥 Safari
- [ ] mobile (iOS/iPadOS)
- [ ] mobile (Android)
- [X] not applicable
- [ ] other (enter in the "Additional environment details" area below)

### Additional environment details

- SharePoint Online
- SharePoint Server 2016
- SharePoint Server 2019
- SharePoint Server Subscription Edition
- All APIs and customization

### Issue description

As a SharePoint consultant, I often must choose the best technology for customization and automation.

Unfortunately, over time, many technologies have emerged and gone again. Also, many new technologies are SharePoint Online exclusive or have limited OnPremise capabilities and documentation of supported platforms is sometimes outdated and / or only targeting SharePoint Online. Please note, there are still many customers that are using SharePoint Server for example because of data privacy or legal reasons.

Therefore, an overview of all currently supported and deprecated extension points and APIs as orientation is needed.

As far as I know, this is the status quo:

| Type | Name | Status | Technology | SharePoint Online compatible | SharePoint Server 2016 / 2019 compatible | SharePoint Server Subscription Edition compatible | Notes |
| - | - | - | - | - | - | - | - |
| API | [Graph](https://learn.microsoft.com/en-us/graph/api/resources/sharepoint?view=graph-rest-1.0) | Maintained (Preferred) | REST, SDKs for: [.NET](https://learn.microsoft.com/en-us/graph/sdks/sdk-installation#install-the-microsoft-graph-net-sdk), [Go](https://learn.microsoft.com/en-us/graph/sdks/sdk-installation#install-the-microsoft-graph-go-sdk), [Java](https://learn.microsoft.com/en-us/graph/sdks/sdk-installation#install-the-microsoft-graph-java-sdk), [JavaScript](https://learn.microsoft.com/en-us/graph/sdks/sdk-installation#install-the-microsoft-graph-javascript-sdk), [PHP](https://learn.microsoft.com/en-us/graph/sdks/sdk-installation#install-the-microsoft-graph-php-sdk), [PowerShell](https://learn.microsoft.com/en-us/graph/sdks/sdk-installation#install-the-microsoft-graph-powershell-sdk), [Python (preview)](https://learn.microsoft.com/en-us/graph/sdks/sdk-installation#install-the-microsoft-graph-python-sdk-preview) | Yes | No | No |
| API | [(2013) REST / Odata](https://learn.microsoft.com/en-us/sharepoint/dev/sp-add-ins/get-to-know-the-sharepoint-rest-service?tabs=csom) | Maintained | REST | Yes | Yes | Yes |
| API | [(2010) REST](https://learn.microsoft.com/en-us/previous-versions/office/developer/sharepoint-2010/ff521587(v=office.14)) | Maintained | REST | Yes | Yes | Yes |
| API | [Collaborative Application Markup Language (CAML)](https://learn.microsoft.com/en-us/sharepoint/dev/schema/collaborative-application-markup-language-caml-schemas) | Maintained  | REST | Yes | Yes | Yes |
| API | [PnPjs 3+](https://pnp.github.io/pnpjs/getting-started/) | Maintained | JavaScript | Yes | No | No |
| API | [PnPjs 2](https://pnp.github.io/pnpjs/v2/SPFx-on-premises/) | [Not Maintained anymore](https://github.com/pnp/pnpjs/blob/version-2/CHANGELOG.md#2023-feb-10) | JavaScript | Yes | Yes | Yes |
| API | [JavaScript Object Model (JSOM)](https://learn.microsoft.com/en-us/sharepoint/dev/spfx/web-parts/guidance/connect-to-sharepoint-using-jsom) | Deprecated | JavaScript, REST | Yes | Yes | Yes |
| API | [PnP PowerShell 2](https://pnp.github.io/powershell/) | Maintained | PowerShell 7, .NET 6 | Yes | No | No |
| API | [PnP PowerShell](https://github.com/pnp/pnp-powershell) | Not maintained anymore | PowerShell 5, .NET Framework 4.6 | Yes | Yes | Yes |
| API | [SharePoint Server Cmdlets](https://learn.microsoft.com/en-us/powershell/sharepoint/sharepoint-server/sharepoint-server-cmdlets) | Maintained | PowerShell 5, .NET Framework 4 | No | Yes | Yes |
| API | [Client Side Object Model (CSOM) .NET Framework version](https://learn.microsoft.com/en-us/sharepoint/dev/sp-add-ins/complete-basic-operations-using-sharepoint-client-library-code) | Maintained | .NET Framework 4 | Yes | Yes | ? | [Missing nuget package for Subscription Edition](https://www.nuget.org/profiles/OfficeDeveloperPlatformTeam) |
| API | [Client Side Object Model (CSOM) .NET Standard version](https://learn.microsoft.com/en-us/sharepoint/dev/sp-add-ins/complete-basic-operations-using-sharepoint-client-library-code) | Maintained | .NET | Yes | No | No |
| API | [Microsoft.Data.OData (V1-V3) NuGet package](https://www.nuget.org/packages/Microsoft.Data.OData) | Deprecated, contains critical bugs | .NET Framework 4, .NET Standard 1.1+, C# | Yes | Yes | Yes | SharePoint Server only supports the old OData V3 standard. Other 3rd Party OData V3 Clients may work. |
| API | [Windows Phone assemblies](https://learn.microsoft.com/en-us/sharepoint/dev/general-development/build-windows-phone-apps-that-access-sharepoint) | Not maintained anymore | .NET Framework, C# | ? | ? | ? |
| API / Customization | [Silverlight redistributable assemblies](https://learn.microsoft.com/en-us/previous-versions/office/developer/sharepoint-2010/ee538971(v=office.14)) | Not maintained anymore | C# | No | No | No |
| Automation | [Power Automate](https://learn.microsoft.com/en-us/power-automate/getting-started) | Maintained (Preferred) | - | Yes | Yes | Yes | Mailing requires Exchange Online. Premium required for server access. |
| Automation | [Azure Logic Apps](https://learn.microsoft.com/en-us/azure/logic-apps/) | Maintained | - | Yes | Yes | Yes |
| Automation | [Webhooks](https://learn.microsoft.com/en-us/sharepoint/dev/apis/webhooks/get-started-webhooks) | Maintained | - | Yes | No | No |
| Automation | [SharePoint 2010 workflow](https://support.microsoft.com/en-us/office/introduction-to-sharepoint-workflow-07982276-54e8-4e17-8699-5056eff4d9e3) | [Deprecated](https://support.microsoft.com/en-us/office/sharepoint-2010-workflow-retirement-1ca3fff8-9985-410a-85aa-8120f626965f) | - | No | Yes | Yes | Supported until 2026 |
| Automation | [SharePoint 2013 workflow](https://learn.microsoft.com/en-us/sharepoint/governance/getting-started-with-sharepoint-server-workflow) | Maintained | - | [Deprecated](https://support.microsoft.com/en-us/office/sharepoint-2013-workflow-retirement-4613d9cf-69aa-40f7-b6bf-6e7831c9691e) | Yes | Yes | In SharePoint Online supported until 2026. Starting with Visual Studio 2022 workflow templates have been removed from Add-In projects. In server environments the workflow runtime (Service Bus and Microsoft Workflow Manager) has been replaced by [SharePoint Workflow Manager](https://techcommunity.microsoft.com/t5/microsoft-sharepoint-blog/announcing-the-release-of-sharepoint-workflow-manager-for/ba-p/3744881) |
| Customization | [SharePoint Framework (SPFx)](https://learn.microsoft.com/en-us/sharepoint/dev/spfx/sharepoint-framework-overview) | Maintained (Preferred) | JavaScript/ TypeScript | Yes | Yes | Yes | OnPremise versions are very old and support only a few extension points. For example: Full page is not supported. |
| Customization | [Power Apps](https://learn.microsoft.com/en-us/power-apps/maker/canvas-apps/customize-list-form) | Maintained (Preferred) | - | Yes | No | No |
| Customization | [SharePoint Add-ins](https://learn.microsoft.com/en-us/sharepoint/dev/sp-add-ins/sharepoint-add-ins) | Maintained | JavaScript/ TypeScript | Deprecated | Yes | Yes  | [M365 and public marketplace retirement starts November 2024 for new and April 2026 for existing tenants, Server remains supported](https://learn.microsoft.com/en-us/sharepoint/dev/sp-add-ins/retirement-announcement-for-add-ins). Should be migrated to SPFx. No App-Parts in OnPremise modern experience sites, App Web not supported for example by migration tools.  |
| Customization | [Farm Solutions](https://learn.microsoft.com/en-us/sharepoint/dev/general-development/build-farm-solutions-in-sharepoint) | Maintained | C# | No | Yes | Yes | Should be used for customizations of SharePoint administrative functions and not for primary use by end users |
| Customization | [No code Sandbox Solutions](https://learn.microsoft.com/en-us/previous-versions/office/developer/sharepoint-2010/ee536577(v=office.14)) | Maintained | JavaScript | No | Yes | Yes | Should be migrated to Add-Ins |
| Customization | [Code-Based Sandbox Solutions](https://learn.microsoft.com/en-us/previous-versions/office/developer/sharepoint-2010/ee536577(v=office.14)) | Removed | C# | No | No | No | Should be migrated to Add-Ins |
| Customization | [SharePoint Designer 2013](https://support.microsoft.com/en-us/office/introducing-sharepoint-designer-66bf58fe-daeb-4fa6-ae84-fd600e0005c1) | [Deprecated](https://support.microsoft.com/en-us/office/sharepoint-2010-workflow-retirement-1ca3fff8-9985-410a-85aa-8120f626965f) | - | Yes | Yes | Yes | Supported until 2026 |
| Customization | [InfoPath Forms Services](https://learn.microsoft.com/en-us/office/client-developer/infopath/infopath-home) | Deprecated | - | No | Yes | Yes | Supported until 2026, should be migrated to Power Apps |

Recent changes:
- 2024-05-03: Clarified, that Add-In and 2013 workflow support in OnPremise environments is not deprecated.

Now to my questions:
1. Can this table be validated, added to the official documentation, and maintained by Microsoft?
2. For SharePoint Online, it is noticeably clear what should be used, but what are the recommended future proof APIs and customization for SharePoint OnPremise development? Unfortunately, the SPFx version is so old, no modern JavaScript / React library can be used, and many features are missing. For offline automation I think the 2013 REST API should be used. Because CSOM / .NET Framework 4.x is old, many new language and .NET features like dependency injection cannot be used.

**Comment by ghost** [comment]
Thank you for reporting this issue. We will be triaging your incoming issue as soon as possible.

**Comment by nick-pape** [comment]
Hi @ChristianGalla , this is a really good post.

1) I'll keep this ticket open for now, as we may either convert this to a pinned discussion or add a page to documentation with this table. Unfortunately, we don't necessarily have the resources to validate it and keep it fully up-to-date, so it would need to be community maintained.

2) I'm working on an answer for this and will update the thread when I have one.

**Comment by fowl2** [comment]
Just a drive by, but the note that CSOM has been updated for .NET Core - see [Using CSOM for .NET Standard instead of CSOM for .NET Framework](https://learn.microsoft.com/en-us/sharepoint/dev/sp-add-ins/using-csom-for-dotnet-standard). See also [pnpcore](https://github.com/pnp/pnpcore).

I definitely agree that it's lack of good up to date(!) guidance is perennial problem with a sprawling and, uh, well historied, platform like SharePoint.
